From c2e4063c07535df1365464c804158ab8c4f87837 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:39 -0400
Subject: [ethernet] i40e: fix fdir programming

Message-id: <1409658042-22689-219-git-send-email-sassmann@redhat.com>
Patchwork-id: 90416
O-Subject: [RHEL7.1 PATCH 218/281] i40e: fix fdir programming
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jesse Brandeburg <jesse.brandeburg@intel.com>

There were a couple of fields in the fdir descriptor setup that
were not being reprogrammed, which left the opportunity for stale
data to be pushed as part of the descriptor next time it was used.

Change-ID: Ieee5c96a7d4713d469693f086c4854de949a7633
Signed-off-by: Jesse Brandeburg <jesse.brandeburg@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 99753ea60626a4c5059b24e2a4c22f8fd88b8207)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_txrx.c b/drivers/net/ethernet/intel/i40e/i40e_txrx.c
index c5749d6..5cc27fb 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_txrx.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_txrx.c
@@ -100,8 +100,6 @@ int i40e_program_fdir_filter(struct i40e_fdir_filter *fdir_data, u8 *raw_packet,
    I40E_TXD_FLTR_QW0_DEST_VSI_SHIFT) &
          I40E_TXD_FLTR_QW0_DEST_VSI_MASK;
 
- fdir_desc->qindex_flex_ptype_vsi = cpu_to_le32(fpt);
-
  dcc = I40E_TX_DESC_DTYPE_FILTER_PROG;
 
  if (add)
@@ -124,6 +122,8 @@ int i40e_program_fdir_filter(struct i40e_fdir_filter *fdir_data, u8 *raw_packet,
    I40E_TXD_FLTR_QW1_CNTINDEX_MASK;
  }
 
+ fdir_desc->qindex_flex_ptype_vsi = cpu_to_le32(fpt);
+ fdir_desc->rsvd = cpu_to_le32(0);
  fdir_desc->dtype_cmd_cntindex = cpu_to_le32(dcc);
  fdir_desc->fd_id = cpu_to_le32(fdir_data->fd_id);
 
@@ -1688,7 +1688,9 @@ static void i40e_atr(struct i40e_ring *tx_ring, struct sk_buff *skb,
   I40E_TXD_FLTR_QW1_CNTINDEX_MASK;
 
  fdir_desc->qindex_flex_ptype_vsi = cpu_to_le32(flex_ptype);
+ fdir_desc->rsvd = cpu_to_le32(0);
  fdir_desc->dtype_cmd_cntindex = cpu_to_le32(dtype_cmd);
+ fdir_desc->fd_id = cpu_to_le32(0);
 }
 
 /**
-- 
1.7.1